Hw03 annotations

vv16vv opened the pull request.

vv16vv committed:
HW03 - annotations: added own Test, After and Before annotations and their Runner.

vv16vv committed:
Main: read tested class from command line arguments

kataus commented on hw03-annotations/src/main/java/ru/otus/vsh/hw03/runner/Runner.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    private static class TestExecutionData {
+        private final Method testMethod;
```
Спорное выделение класса внутри класса для группировки полей... У нас и так всё инкапсулировано, это только усложняет и сбивает с толку. Легко разбивается на внутренние поля и метод, который их правильно заполнит.

kataus commented on hw03-annotations/src/main/java/ru/otus/vsh/hw03/runner/Runner.java:
```diff
@@ -0,0 +1,138 @@
+            if (hasPassed) passed++;
+            else failed++;
+        }
+        return Map.of(Status.SUCCESS, passed, Status.FAILED, failed);
```
Вот смотрите, опять же вопрос с инкапсуляцией. У нас на каждый тестовый класс запускается конкретный экземпляр Runner. Соответственно мы можем безопасно использовать  passed и failed как приватные поля и не городить такую сложную конструкцию.
Или если хочется оставить TestExecutionData  - то можно сделать внутри него поля passed и failed и там установить.

kataus commented on hw03-annotations/src/main/java/ru/otus/vsh/hw03/runner/Runner.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Nonnull
+    public static Map<Status, Integer> run(String className) throws ClassNotFoundException {
```
А зачем так усложнять статикой? Если мы хотим чтобы был прям статичный запуск, то давайте в нем оставим только инициализацию Runner и запуск инициированного runner в работу.

kataus replied:
Принято придерживаться следующего параметра записи методов класса по ходу: constructor, public, default, protected, private. А то сейчас ключевой метод в классе оказался в подвале текста, пока до него долистаешь в IDE

vv16vv committed:
ComplexTest renamed to WithTimesTest as it correcter represent test's goal.

vv16vv committed:
SeveralTestsTest: added test with several test methods.

vv16vv committed:
Changes according to PR comments

Согласна с Вашим замечанием и Runner стал обычным классом. Вложенный класс переименован в TestInfo, что лучше отражает цель класса - хранение вспомогательной информации для каждого тестового метода. Так как тестовых методов может быть несколько, то если выносить эти параметры в класс Runner - пришлось бы делать какие-то дополнительные массивы или мапы по одному на каждый параметр, что на мой взгляд не очень красиво. Общие данные - списки методов Before и After - вынесены, они не меняются от одного тестового метода к другому. Метод executeTest переименован в executeTestSet, чтобы подчеркнуть, что исполняется не один тестовый метод, а целый набор методов. Счетчики внутри этого метода я оставила локальными - мне кажется, что они это внутренняя реализация подсчета статистики, а далее на выход идет упорядоченная мапа (по сути мне просто нужен тут массив где индексом является статус). Мне хотелось бы тут отделить место где вычисляется статистика от мест где используются результаты.

kataus commented on hw03-annotations/src/main/java/ru/otus/vsh/hw03/runner/Runner.java:
```diff
@@ -0,0 +1,159 @@
+        aClass = Class.forName(className);
+        beforeMethods = getAnnotatedMethods(Before.class);
+        afterMethods = getAnnotatedMethods(After.class);
+        testMethods = getAnnotatedMethods(Test.class).stream().map(TestInfo::new).collect(Collectors.toList());
```
Ни разу не видел тут использование конструкторов. Но корректно и неплохо. Из советов - почитайте на досуге Джошуа Блоха Effective Java. Там он разносит конструкторы в пух и прах. Оно не горит и не критично, но рекомендую для общего развития

vv16vv replied:
Вы о том, что эксепшен выбрасывается из конструктора? Да, мне тоже не очень нравится. Переделала немного по-другому - теперь без эксепшена, если не ошибаюсь, теперь используется template method.

kataus commented on hw03-annotations/src/main/java/ru/otus/vsh/hw03/runner/Runner.java:
```diff
@@ -0,0 +1,159 @@
+        return Map.of(TestStatus.SUCCESS, passed, TestStatus.FAILED, failed);
+    }
+
+    private void initializeResults() {
```
тут всё же скорее drop, а не init :) Но может это моё видение

vv16vv replied:
ммм. Согласна :) только drop больше с базами данных ассоциируется. Переименовала в resetResults.

kataus commented on hw03-annotations/src/main/java/ru/otus/vsh/hw03/runner/Runner.java:
```diff
@@ -0,0 +1,159 @@
+        } catch (Exception e) {
+            throw new IllegalStateException("Error on creating an object of type " + aClass.getSimpleName(), e);
+        }
+        for (int i = 0; i < info.times; i++) {
```
Круто сделано

vv16vv committed:
Use template method to create object of Runner

(avoid of throwing an exception from the constructor)

vv16vv committed:
Renamed initializeResults to ResetResults
